backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/ReflectionSupport.kt

```diff
@@ -10,63 +10,140 @@ import org.jetbrains.kotlin.backend.konan.KonanBackendContext
 import org.jetbrains.kotlin.ir.util.constructors
+import org.jetbrains.kotlin.ir.util.fqNameForIrSerialization
 import org.jetbrains.kotlin.ir.util.irCall
+import org.jetbrains.kotlin.ir.util.render
 import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameUnsafe
 import org.jetbrains.kotlin.resolve.descriptorUtil.getAllSuperClassifiers
+import org.jetbrains.kotlin.types.Variance
 
-internal class KTypeGenerator(
-        private val context: KonanBackendContext,
-        private val eraseTypeParameters: Boolean = false
-) {
+internal class KTypeGenerator(private val context: KonanBackendContext) {
     private val symbols = context.ir.symbols
 
-    fun IrBuilderWithScope.irKType(type: IrType): IrExpression = if (type !is IrSimpleType) {
+    fun IrBuilderWithScope.irKType(type: IrType, needExactTypeParameters: Boolean = false): IrExpression {
+        return irKType(type, needExactTypeParameters, mutableSetOf())
+    }
+
+    private class RecursiveBoundsException(message: String) : Throwable(message)
+
+    private fun IrBuilderWithScope.irKType(
+            type: IrType,
+            needExactTypeParameters: Boolean,
+            seenTypeParameters: MutableSet<IrTypeParameter>,
+    ): IrExpression = if (type !is IrSimpleType) {
         // Represent as non-denotable type:
         irKTypeImpl(
                 kClassifier = irNull(),
                 irTypeArguments = emptyList(),
-                isMarkedNullable = false
+                isMarkedNullable = false,
+                needExactTypeParameters = needExactTypeParameters,
+                seenTypeParameters = seenTypeParameters
         )
-    } else {
-        val classifier = type.classifier
-
-        if (classifier is IrClassSymbol) {
-            irKTypeImpl(
-                    kClassifier = irKClass(classifier),
-                    irTypeArguments = type.arguments,
-                    isMarkedNullable = type.hasQuestionMark
-            )
-        } else {
-            if (eraseTypeParameters) {
-                irKType(context.irBuiltIns.anyNType)
-            } else {
-                irCall(symbols.kTypeImplForGenerics.constructors.single())
-            }
+    } else try {
+        val kClassifier = when (val classifier = type.classifier) {
+            is IrClassSymbol -> irKClass(classifier)
+            is IrTypeParameterSymbol -> irKTypeParameter(classifier.owner, needExactTypeParameters, seenTypeParameters)
+            else -> TODO("Unexpected classifier: $classifier")
         }
+
+        irKTypeImpl(
+                kClassifier = kClassifier,
+                irTypeArguments = type.arguments,
+                isMarkedNullable = type.hasQuestionMark,
+                needExactTypeParameters = needExactTypeParameters,
+                seenTypeParameters = seenTypeParameters
+        )
+    } catch (t: RecursiveBoundsException) {
+        if (needExactTypeParameters)
+            this@KTypeGenerator.context.reportCompilationError(t.message!!)
```

Comment:
It should report this error with source location.
Maybe introduce a class encapsulating all generation parameters and state?

Reply:
Done.